### DISCUSSION TITLE
Using rustfmt during yocto build

### DISCUSSION THREAD

**Discussion by kevswims** [meta-rust/meta-rust#350] [Q&A] [answered]
Is there a way to build a rust package as a `native` package for use during the build? Our project relies on `tonic_build` https://docs.rs/tonic-build/0.4.2/tonic_build/ which by default calls `rustfmt` on some generated code. We have had to add extra logic to disable that step if `rustfmt` is not available. 

From my experience with other yocto packages I would think this should be something like adding a `DEPENDS += "rustfmt-native"` but that does not work.

Am I missing the way to do this or should I open an issue to request this?

**Comment by weltling** [comment, marked as answer]
There's currently no `rustfmt` on the native side but also it doesn't seem to make sense to have that during the image build. As why would one format some code anyway before compiling it, the only outcome of interest would be the resulting binaries. Also, to produce the tool itself it'd take additional build time as all the toolchain is produced anew (except sstate cache is used).

There's #345 targeting `rustfmt` availability in the SDK, though. That's where being able to format code would be useful.

Thanks

**Reply by kevswims** [reply]
That makes sense. It's probably not worth doing this then since the only benefit would be a slight simplification of our build scripts and I cannot come up with any other reason why I would want to do this.

**Reply by bernardoaraujor** [reply]
@kevswims I'm in a similar situation that you described on your post.
What was your solution? Did you fork `tonic_build` and disable `rustfmt`?

**Reply by kevswims** [reply]
@bernardoaraujor In our build.rs we do a check to see if `rustfmt` is available like this:

```
    let fmt_available = Command::new("rustfmt").arg("--version").output().is_ok();
```

Then when we call `tonic_build::configure` we pass that in to turn formatting on or off like this:
```
    tonic_build::configure().format(fmt_available).compile(
        ...
    )?;
```

**Reply by bernardoaraujor** [reply]
hmm thanks @kevswims !

I'm just patching `Cargo.toml` to disable `codegen-rustfmt` as a feature.

```
 [features]
-default = ["codegen-rustfmt"]
 with-bench = []
-codegen-rustfmt = ["chain-network/codegen-rustfmt"]
 integration-test = []
 soak-test = []
 systemd = ["tracing-journald"]
```

Perhaps your approach is more elegant, but in my case there's a few degrees of dependency between the package I'm actually bitbaking ([`jormungandr`](https://github.com/input-output-hk/jormungandr)), and crate that uses `tonic_build` ([`chain-libs/chain-network`](https://github.com/input-output-hk/chain-libs/blob/715644b517be119eb856ca06893e01edbd777995/chain-network/Cargo.toml#L35)).

So patching `jormungandr`'s `Cargo.toml` felt a like a shorter path. Also I don't need any forks.

**Reply by kevswims** [reply]
Ahh, that is more difficult since it is not within your build.rs. We are the one's using protobuf directly in our code so it was easy to put this into the build.rs. 

I wonder if this would be better solved by an environment variable that would disable formatting? That would be easy to set in the yocto recipe at least.
